clients/src/main/java/org/apache/kafka/common/memory/RecyclingMemoryPool.java

```diff
@@ -0,0 +1,142 @@
+            throw new IllegalArgumentException(String.format("Must provide a positive cacheable buffer size and buffer cache " +
+                    "capacity, provided %d and %d respectively.", cacheableBufferSize, bufferCacheCapacity));
+        }
+        this.bufferCache = new LinkedList<>();
+        this.cacheableBufferSize = cacheableBufferSize;
+        this.bufferCacheCapacity = bufferCacheCapacity;
+        this.requestSensor = requestSensor;
+        this.bufferCacheSlot = 0;
+    }
+
+    @Override
+    public ByteBuffer tryAllocate(int sizeBytes) {
+        if (sizeBytes < 1) {
+            throw new IllegalArgumentException("requested size " + sizeBytes + "<=0");
+        }
+
+        ByteBuffer allocated = null;
+        if (sizeBytes > cacheableBufferSize / 2 && sizeBytes <= cacheableBufferSize) {
+            allocated = maybePopBufferFromCache();
+        }
+        if (allocated == null) {
+            allocated = ByteBuffer.allocate(sizeBytes);
+        }
+        bufferToBeAllocated(allocated);
+        return allocated;
+    }
+
+    /**
+     * If (1) there are recycled buffers in cache or (2) more buffer of size {@link #cacheableBufferSize} can be allocated,
+     * (allocate the buffer and) return the buffer to the client.
+     *
+     * @return The available buffer if any.
+     */
+    protected ByteBuffer maybePopBufferFromCache() {
+        boolean canAllocateMoreBuffer;
+        synchronized (this) {
+            if (!bufferCache.isEmpty()) {
+                Iterator<ByteBuffer> it = bufferCache.iterator();
+                ByteBuffer buffer = it.next();
+                it.remove();
+                buffer.clear();
+                return buffer;
+            }
+            canAllocateMoreBuffer = bufferCacheSlot < bufferCacheCapacity;
+            if (canAllocateMoreBuffer) {
+                bufferCacheSlot++;
+            }
+        }
+        if (canAllocateMoreBuffer) {
+            ByteBuffer byteBuffer = ByteBuffer.allocate(cacheableBufferSize);
+            bufferToBeAllocated(byteBuffer);
+            return byteBuffer;
+        }
+        return null;
+    }
+
+    @Override
+    public void release(ByteBuffer previouslyAllocated) {
+        if (previouslyAllocated == null) {
+            throw new IllegalArgumentException("provided null buffer");
+        }
+        if (previouslyAllocated.capacity() == cacheableBufferSize) {
+            maybeRecycleBufferToCache(previouslyAllocated);
+        }
+        bufferToBeReleased(previouslyAllocated);
+    }
+
+    protected synchronized void maybeRecycleBufferToCache(ByteBuffer previouslyAllocated) {
+        if (bufferCache.size() < bufferCacheCapacity) {
+            bufferCache.add(previouslyAllocated);
+        }
+    }
+
+    //allows subclasses to do their own bookkeeping (and validation) _before_ memory is returned to client code.
+    protected void bufferToBeAllocated(ByteBuffer justAllocated) {
```

Comment:
I am not sure if it is the case that the thread which allocates some memory will always be the thread that releases that memory. If the it is always the same thread that allocates and releases the same memory, how about let's log the thread ID as well? So we can get a better insights from the log

Reply:
This memory pool is shared by all the socket channel (that why it has to be thread safe).